### PR TITLE
BaSP-TG-38: add edit time sheet page

### PR DESCRIPTION
**Pull request by Licamus**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -12,7 +12,7 @@
     "web-vitals": "^1.1.2"
   },
   "scripts": {
-    "start": "react-scripts start",
+    "start": "react-scripts --openssl-legacy-provider start",
```
Comment (lmalvasia) [inline package.json:15]:
```suggestion
    "start": "react-scripts start",
```

File: src/Components/TimeSheets/Form/Form Modal/index.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import styles from './modal.module.css';
+
+function Modal(props) {
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/Form Modal/index.js:3]:
```suggestion
function FormModal(props) {
```

Comment (lmalvasia) [reply src/Components/TimeSheets/Form/Form Modal/index.js:3]: Rename the folder to FormModal (without spaces)

File: src/Components/TimeSheets/Form/Form Title/index.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import styles from './formTitle.module.css';
+
+function FormTitle(props) {
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/Form Title/index.js:3]: Remove this FormTitle component. It's not necessary to have a component just to render an h3.

File: src/Components/TimeSheets/Form/index.js [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+    description: '',
+    date: '',
+    hours: '',
+    task: '',
+    employee: '',
+    project: ''
+  });
+  const [showModal, setShowModal] = useState(false);
+  const [serverError, setServerError] = useState('');
+  const [formMode, setFormMode] = useState(true);
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:17]:
```suggestion
  const [isEditing, setIsEditing] = useState(false);
```

File: src/Components/TimeSheets/Form/index.js [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+  const [projects, setProjects] = useState([]);
+  const [employees, setEmployees] = useState([]);
+  const [employeesTotal, setEmployeesTotal] = useState([]);
+
+  const onChangeInputValue = (e) => {
+    setInputTimeSheetValue({ ...inputTimeSheetValue, [e.target.name]: e.target.value });
+
+    if (e.target.name === 'project') {
+      const selectedProject = projects.find((project) => project._id === e.target.value);
+      console.log(selectedProject);
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:29]: Remove the console log

File: src/Components/TimeSheets/Form/index.js [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+      const projectEmployees = selectedProject.employees.map((employee) => employee.employee);
+      console.log(projectEmployees);
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:31]: Remove this console log

File: src/Components/TimeSheets/Form/index.js [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+      setEmployees(projectEmployees);
+    }
+  };
+
+  useEffect(async () => {
+    try {
+      const response2 = await fetch(`${process.env.REACT_APP_API_URL}/tasks`, {
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:38]:
```suggestion
      const response = await fetch(`${process.env.REACT_APP_API_URL}/tasks`, {
```

File: src/Components/TimeSheets/Form/index.js [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+        method: 'GET'
+      });
+      const json = await response2.json();
+      setTasks(json.data);
+    } catch (error) {
+      console.error(error);
+    }
+
+    try {
+      const response4 = await fetch(`${process.env.REACT_APP_API_URL}/employees`, {
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:48]:
```suggestion
      const response = await fetch(`${process.env.REACT_APP_API_URL}/employees`, {
```

File: src/Components/TimeSheets/Form/index.js [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+        method: 'GET'
+      });
+      const json = await response4.json();
+      setEmployeesTotal(json.data);
+    } catch (error) {
+      console.error(error);
+    }
+
+    try {
+      const response3 = await fetch(`${process.env.REACT_APP_API_URL}/projects`, {
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:58]:
```suggestion
      const response = await fetch(`${process.env.REACT_APP_API_URL}/projects`, {
```

File: src/Components/TimeSheets/Form/index.js [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+        method: 'GET'
+      });
+      const json = await response3.json();
+      setProjects(json.data);
+    } catch (error) {
+      console.error(error);
+    }
+
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:66]:
```suggestion
      const params = new URLSearchParams(window.location.search);
      const id = params.get('id');
```

File: src/Components/TimeSheets/Form/index.js [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+    if (window.location.href.includes('id=')) {
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:67]:
```suggestion
    if (id) {
```

File: src/Components/TimeSheets/Form/index.js [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+      try {
+        const url = window.location.href;
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:69]: Remove this

File: src/Components/TimeSheets/Form/index.js [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+        const id = url.substring(url.lastIndexOf('=') + 1);
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:70]: Remove this

File: src/Components/TimeSheets/Form/index.js [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+        const response = await fetch(`${process.env.REACT_APP_API_URL}/time-sheets/${id}`, {
+          method: 'GET'
+        });
+        const json = await response.json();
+        setFormMode(false);
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:75]:
```suggestion
        setIsEditing(true);
```

File: src/Components/TimeSheets/Form/index.js [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+        setFormTitle('Update TimeSheet');
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:76]: Remove this.

File: src/Components/TimeSheets/Form/index.js [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+    date: '',
+    hours: '',
+    task: '',
+    employee: '',
+    project: ''
+  });
+  const [showModal, setShowModal] = useState(false);
+  const [serverError, setServerError] = useState('');
+  const [formMode, setFormMode] = useState(true);
+  const [formTitle, setFormTitle] = useState('Add timeSheet');
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:18]: This is not necessary, we could remove it.

File: src/Components/TimeSheets/Form/index.js [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+          date: correctDate(json.data.date),
+          hours: json.data.hours,
+          task: json.data.task['_id'],
+          employee: json.data.employee['_id'],
+          project: json.data.project['_id']
+        });
+      } catch (error) {
+        console.error(error);
+      }
+    } else {
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:88]: This else statement is not necessary. Remove it.

File: src/Components/TimeSheets/Form/index.js [hunk 18]
```diff
@@ -0,0 +1,280 @@
...
+    let dateFormated = date.substr(0, 10);
+    return dateFormated;
+  };
+
+  const closeModal = () => {
+    setShowModal(false);
+  };
+
+  const onSubmit = async (event) => {
+    if (formMode) {
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:103]:
```suggestion
    if (!isEditing) {
```

File: src/Components/TimeSheets/Form/index.js [hunk 19]
```diff
@@ -0,0 +1,280 @@
...
+      event.preventDefault();
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:104]: Move the prevent default to the top of this function.

File: src/Components/TimeSheets/Form/index.js [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+      });
+      const content = await rawResponse.json();
+      if (!content.error) {
+        window.location.assign('/time-sheets');
+      } else {
+        setShowModal(true);
+        setServerError(content.message);
+      }
+    } else {
+      event.preventDefault();
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:128]: Remove this one.

File: src/Components/TimeSheets/Form/index.js [hunk 21]
```diff
@@ -0,0 +1,280 @@
...
+      const url = window.location.href;
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:129]: Use the URLSearchParams function to get this.
      const params = new URLSearchParams(window.location.search);
      const id = params.get('id');

File: src/Components/TimeSheets/Form/index.js [hunk 22]
```diff
@@ -0,0 +1,280 @@
...
+    }
+  };
+
+  return (
+    <div className={styles.container}>
+      <Modal show={showModal} title={serverError} closeModal={closeModal} />
+      <form onSubmit={onSubmit}>
+        <div>
+          <div className={styles.cardTitle}>
+            <FormTitle title={formTitle} />
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:162]: Remove this and replace with the h3 directly.
You can do:
`<h3>{isEditing ? 'Edit time sheet' : 'Create timesheet}</h3>`

File: src/Components/TimeSheets/Form/index.js [hunk 23]
```diff
@@ -0,0 +1,280 @@
...
+              name="employee"
+              value={inputTimeSheetValue.employee}
+              onChange={onChangeInputValue}
+              required
+            >
+              <option value="" disabled hidden>
+                Select an employee
+              </option>
+              {employees.map((employee, index) => {
+                console.log(employees);
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:226]: Remove all the following console logs

File: src/Components/TimeSheets/TimeSheet/index.js [hunk 24]
```diff
@@ -0,0 +1,70 @@
+import Trash from '../../../assets/trash.png';
+import styles from './time.module.css';
+
+const TimeSheet = ({ timeSheet, setShowModal, setTimeSheetId }) => {
+  return (
+    <tr style={{ display: 'flex', justifyContent: 'center' }}>
+      <td
+        onClick={() => {
```
Comment (lmalvasia) [inline src/Components/TimeSheets/TimeSheet/index.js:8]: The onClick function should be in the `<tr>`

File: src/Components/TimeSheets/TimeSheet/index.js [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+        style={{ flexBasis: '10%' }}
+      >
+        {timeSheet.project.name}
+      </td>
+      <td style={{ flexBasis: '10%' }}>
+        <img
+          className={styles.button}
+          src={Trash}
+          alt="trash"
+          onClick={() => {
```
Comment (lmalvasia) [inline src/Components/TimeSheets/TimeSheet/index.js:60]:
```suggestion
          onClick={(event) => {
          event.stopPropagation();
```

File: package.json (hunk 1 above)
Comment (Licamus) [reply package.json:15]: Done!

File: src/Components/TimeSheets/Form/Form Modal/index.js (hunk 2 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/Form Modal/index.js:3]: Done!

File: src/Components/TimeSheets/Form/index.js (hunk 4 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:17]: Done!

File: src/Components/TimeSheets/Form/index.js (hunk 5 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:29]: Done!

File: src/Components/TimeSheets/Form/index.js (hunk 6 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:31]: Done!

File: src/Components/TimeSheets/Form/index.js (hunk 7 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:38]: done!

File: src/Components/TimeSheets/Form/index.js (hunk 8 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:48]: done!

File: src/Components/TimeSheets/Form/index.js (hunk 9 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:58]: done!

File: src/Components/TimeSheets/Form/index.js (hunk 18 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:103]: done!

File: src/Components/TimeSheets/Form/index.js (hunk 19 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:104]: done!

File: src/Components/TimeSheets/Form/index.js (hunk 20 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:128]: done!

File: src/Components/TimeSheets/Form/index.js (hunk 23 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:226]: Done!

File: src/Components/TimeSheets/TimeSheet/index.js (hunk 25 above)
Comment (Licamus) [reply src/Components/TimeSheets/TimeSheet/index.js:60]: Done

File: src/Components/TimeSheets/TimeSheet/index.js (hunk 24 above)
Comment (Licamus) [reply src/Components/TimeSheets/TimeSheet/index.js:8]: done

File: src/Components/TimeSheets/Form/index.js (hunk 22 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:162]: doesn't work with isEditing, so i left this as it was previously.

Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:162]: done!

File: src/Components/TimeSheets/Form/index.js (hunk 14 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:75]: Done!

File: src/Components/TimeSheets/Form/index.js (hunk 15 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:76]: Done!

File: src/Components/TimeSheets/Form/index.js (hunk 16 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:18]: Done!

File: src/Components/TimeSheets/Form/index.js (hunk 21 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:129]: Done!

File: src/Components/TimeSheets/Form/index.js (hunk 13 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:70]: Done

File: src/Components/TimeSheets/Form/index.js (hunk 12 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:69]: done

File: src/Components/TimeSheets/Form/index.js (hunk 11 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:67]: done

File: src/Components/TimeSheets/Form/index.js (hunk 10 above)
Comment (Licamus) [reply src/Components/TimeSheets/Form/index.js:66]: done

**Review by lmalvasia — APPROVED** [review]

File: src/Components/TimeSheets/Form/index.js [hunk 26]
```diff
@@ -0,0 +1,269 @@
...
+    setInputTimeSheetValue({ ...inputTimeSheetValue, [e.target.name]: e.target.value });
+
+    if (e.target.name === 'project') {
+      const selectedProject = projects.find((project) => project._id === e.target.value);
+      const projectEmployees = selectedProject.employees.map((employee) => employee.employee);
+      setEmployees(projectEmployees);
+    }
+  };
+
+  useEffect(async () => {
```
Comment (lmalvasia) [inline src/Components/TimeSheets/Form/index.js:32]: We should have only one try catch on this useEffect. We also should populate the project employees list. If we not populate this, when we try to edit an employee, the select is empty.
```
  useEffect(async () => {
    try {
      const tasksResponse = await fetch(`${process.env.REACT_APP_API_URL}/tasks`, {
        method: 'GET'
      });
      const tasks = await tasksResponse.json();
      setTasks(tasks.data);
      const employeesResponse = await fetch(`${process.env.REACT_APP_API_URL}/employees`, {
        method: 'GET'
      });
      const employees = await employeesResponse.json();
      setEmployeesTotal(employees.data);
      const projectsResponse = await fetch(`${process.env.REACT_APP_API_URL}/projects`, {
        method: 'GET'
      });
      const projects = await projectsResponse.json();
      setProjects(projects.data);
      const params = new URLSearchParams(window.location.search);
      const id = params.get('id');
      if (id) {
        const timeSheetResponse = await fetch(`${process.env.REACT_APP_API_URL}/time-sheets/${id}`, {
          method: 'GET'
        });
        const timeSheet = await timeSheetResponse.json();
        const selectedProject = projects.find((project) => project._id === timeSheet.project._id);
        const projectEmployees = selectedProject.employees.map((employee) => employee.employee);
        setEmployees(projectEmployees);
        setIsEditing(true);
        setInputTimeSheetValue({
          description: json.data.description,
          date: correctDate(json.data.date),
          hours: json.data.hours,
          task: json.data.task['_id'],
          employee: json.data.employee['_id'],
          project: json.data.project['_id']
        });
      }
    } catch (error) {
      console.error(error);
      alert(error);
    }
  }, []);
```

**Review by lmalvasia — DISMISSED** [review]

**Review by Romimonde — APPROVED** [review]
:)

**Pull request merged**